http/responseemitter.go

```diff
@@ -136,23 +129,43 @@ func (re *responseEmitter) Emit(value interface{}) error {
 }
 
 func (re *responseEmitter) SetLength(l uint64) {
+	re.l.Lock()
+	defer re.l.Unlock()
+
 	h := re.w.Header()
 	h.Set("X-Content-Length", strconv.FormatUint(l, 10))
 
 	re.length = l
 }
 
 func (re *responseEmitter) Close() error {
-	re.once.Do(func() { re.preamble(nil) })
-	return nil
+	return re.CloseWithError(nil)
 }
 
-func (re *responseEmitter) SetError(v interface{}, errType cmdkit.ErrorType) {
-	err := re.Emit(&cmdkit.Error{Message: fmt.Sprint(v), Code: errType})
+func (re *responseEmitter) CloseWithError(err error) error {
+	re.l.Lock()
+	defer re.l.Unlock()
+
+	if re.w == nil {
+		return cmds.ErrClosingClosedEmitter
+	}
+
+	return re.closeWithError(err)
+}
+
+func (re *responseEmitter) closeWithError(err error) error {
+	if err == io.EOF {
+		err = nil
+	}
 	if err != nil {
-		log.Debug("http.SetError err=", err)
-		panic(err)
+		// abort by sending an error trailer
+		re.w.Header().Set(StreamErrHeader, err.Error())
```

Comment:
This is where we'd *also* send an error value if we decided to go that route. Seeing this, I'm now pretty sure we're not reading an error somewhere where we should be (I don't see us *reading* this trailer anywhere).

Reply:
I now check the header errors here: https://github.com/ipfs/go-ipfs-cmds/pull/112/files/#diff-c8a8dcac9ba61ed253376bf559309ae4R90